NewForce-at-Mountwest/bangazon-site-the-grave-trolls
# Users can now checkout

# Description

When viewing an open order, users can now click the Checkout button, then select a Payment Type to finalize their order. Users cannot order items when their order will put the item quantity below 0. Fixed all of the product hyperlinks to make sure they directed to the Products controller.

## Type of change


# Related Tickets

[Ticket 2](https://trello.com/c/JuHphjrN/2-2-user-can-complete-an-order)

# Testing Instructions

- Clone down these changes on a new branch using `git checkout -b db-checkout` and `git pull origin db-checkout`
- Open Visual Studio
- If you don't have an AppSettings.json file, create one in the `Bangazon` folder of the project and copy the following code:
```
{
  "Logging": { "LogLevel": { "Default": "Warning" } },
  "AllowedHosts": "*",
  "ConnectionStrings": { "DefaultConnection": "Server=localhost\\SQLEXPRESS01;Database=BangazonSite;Trusted_Connection=True;" }
}
```
- Adjust the SQLEXPRESS line of AppSettings.json to match the server running on your machine
- Click the IIS Express button (or F5) to run the app
- Try creating a new user, so we can see what our cart looks like without a payment type.
- Add a few products to your cart, then view your cart.
- Try checking out before you've added a payment type. Doing so should give you a message saying you need to add a payment to your account. 
- Add a payment, then try to checkout again. Add multiple payment types if you want! When you checkout again, you'll be given a dropdown of your payment types. Select one and finalize your checkout.
- Click "Sell Something" to add a new product listing. Give it a small quantity, like 1 or 2.
- Add that new product to your cart. Add more than is in stock (if you said there is 1 in stock, add like 3)
- Navigate to your cart. Try adding an `/1` after the URL to see if you can view someone else's order. You should get a 404
- Try this same thing but with an id of 3 to try to view your past order. This should work, even though we haven't linked it anywhere in the app yet.
- Try checking out again. This is technically an Edit view. Try changing the id's in the URL again. It shouldn't work - you should only be able to edit an order with your user id and a null payment type.
- Now that we've tested all the URL nonsense, try checking out!
- After you've chosen a payment type, you'll be directed to the final checkout screen again. This should show you the products you were able to buy, but should ALSO tell you that some items were out of stock and removed.

# Checklist:
